File: scala/scopes.scala
package com.example

// Labeled scope test cases for Scala. See SCOPE_TEST_SPEC.md.

import scala.collection.mutable

// Import with alias — used for S09.
import com.example.scopes_ns.{Widget => NsWidget}    // S09.def

object Scopes {
  var moduleVar: String = "mod-initial"              // S04.def

  def s01Local(): Unit = {
    val localA = "S01.local"                         // S01.def
    println(localA)                                  // S01.read
  }

  def s02ClosureRead(): Unit = {
    val outerA = "S02.outer"                         // S02.outer.def
    val inner = () => println(outerA)                // S02.inner.read
    inner()
  }

  def s03ClosureWrite(): Int = {
    var counter = 0                                  // S03.outer.def
    val bump = () => counter += 1                    // S03.inner.write
    bump(); bump()
    counter                                          // S03.outer.read
  }

  def s05SameModuleWrite(): Unit = {
    moduleVar = "rotated"                            // S05.write
    println(moduleVar)                               // S05.read
  }

  def s08Shadowing(): Unit = {
    val moduleVar = "shadowed"                       // S08.shadow.def
    println(moduleVar)                               // S08.shadow.read
  }

  def s09AliasedImport(): Unit = {
    val w = new NsWidget("via-alias")                // S09.read
    println(w.label)
  }

  class Base(val x: Int) {                           // S11.instance.def
    def readInstance(x: Int): Int = x + this.x       // S11.param.read + S11.instance.read
  }

  object Base {                                      // S12.static.def (companion)
    var staticX: Int = 1
  }

  class Derived extends Base(5) {
    def readInherited: Int = Base.staticX            // S13.derived.read
  }

  def s14Qualified(): String = new NsWidget("hi").label   // S14.read

  def run(): Unit = {
    s01Local()
    s02ClosureRead()
    println(s03ClosureWrite())
    s05SameModuleWrite()
    s08Shadowing()
    s09AliasedImport()
    println(new Base(42).readInstance(100))
    println((new Derived).readInherited)
    println(s14Qualified())
  }
}

File: swift/scopes.swift
// Labeled scope test cases for Swift. See SCOPE_TEST_SPEC.md.

import Foundation

var moduleVar: String = "mod-initial"               // S04.def

func s01Local() {
    let localA = "S01.local"                        // S01.def
    print(localA)                                   // S01.read
}

func s02ClosureRead() {
    let outerA = "S02.outer"                        // S02.outer.def
    let inner = { print(outerA) }                   // S02.inner.read
    inner()
}

func s03ClosureWrite() -> Int {
    var counter = 0                                 // S03.outer.def
    let bump = { counter += 1 }                     // S03.inner.write
    bump(); bump()
    return counter                                  // S03.outer.read
}

func s05SameModuleWrite() {
    moduleVar = "rotated"                           // S05.write
    print(moduleVar)                                // S05.read
}

func s08Shadowing() {
    let moduleVar = "shadowed"                      // S08.shadow.def
    print(moduleVar)                                // S08.shadow.read
}

class ScopeBase {
    static var staticX: Int = 1                     // S12.static.def / S13.base.def
    var x: Int                                      // S11.instance.def
    init(_ x: Int) { self.x = x }
    func readInstance(_ x: Int) -> Int {
        return x + self.x                           // S11.param.read + S11.instance.read
    }
}

class ScopeDerived: ScopeBase {
    init() { super.init(5) }
    func readInherited() -> Int { ScopeBase.staticX }  // S13.derived.read
}

struct Widget {                                     // S14.Widget.def
    let label: String
}

func s14Qualified() -> String { Widget(label: "hi").label } // S14.read

func runScopeDemoSwift() {
    s01Local()
    s02ClosureRead()
    print(s03ClosureWrite())
    s05SameModuleWrite()
    s08Shadowing()
    print(ScopeBase(42).readInstance(100))
    print(ScopeDerived().readInherited())
    print(s14Qualified())
}
